driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java

```diff
@@ -107,30 +111,49 @@ public final class CommandMessage extends RequestMessage {
         this.serverApi = serverApi;
     }
 
+    // Create a BsonDocument representing the logical document encoded by an OP_MSG.
+    // The returned document will contain all the fields, including ones represented by
+    // OP_MSG Sections of type Kind 1: Document Sequence
     BsonDocument getCommandDocument(final ByteBufferBsonOutput bsonOutput) {
-        ByteBufBsonDocument byteBufBsonDocument = ByteBufBsonDocument.createOne(bsonOutput,
-                getEncodingMetadata().getFirstDocumentPosition());
-        BsonDocument commandBsonDocument;
-
-        if (containsPayload()) {
-            commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
-
-            int payloadStartPosition = getEncodingMetadata().getFirstDocumentPosition()
-                    + byteBufBsonDocument.getSizeInBytes()
-                    + 1 // payload type
-                    + 4 // payload size
-                    + payload.getPayloadName().getBytes(StandardCharsets.UTF_8).length + 1;  // null-terminated UTF-8 payload name
-            commandBsonDocument.append(payload.getPayloadName(),
-                    new BsonArray(ByteBufBsonDocument.createList(bsonOutput, payloadStartPosition)));
-        } else {
-            commandBsonDocument = byteBufBsonDocument;
+        List<ByteBuf> duplicateByteBuffers = bsonOutput.getByteBuffers();
+        try {
+            CompositeByteBuf outputByteBuf = new CompositeByteBuf(duplicateByteBuffers);
+            outputByteBuf.position(getEncodingMetadata().getFirstDocumentPosition());
+            ByteBufBsonDocument byteBufBsonDocument = createOne(outputByteBuf);
+
+            if (outputByteBuf.hasRemaining()) {
+                BsonDocument commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
+
+                while (outputByteBuf.hasRemaining()) {
+                    outputByteBuf.position(outputByteBuf.position() + 1 /* payload type */ + 4 /* payload size */);
+                    String payloadName = getPayloadName(outputByteBuf);
+                    assertFalse(payloadName.contains("."));
+                    commandBsonDocument.append(payloadName, new BsonArray(createList(outputByteBuf)));
+                }
+                return commandBsonDocument;
+            } else {
+                return byteBufBsonDocument;
+            }
+        } finally {
+            duplicateByteBuffers.forEach(ByteBuf::release);
         }
-
-       return commandBsonDocument;
     }
 
-    boolean containsPayload() {
-        return payload != null;
+    // Get the field name from a ByteBuf positioned at the start of the document sequence identifier of an OP_MSG Section of type
+    // Kind 1: Document Sequence.
+    // Upon normal completion of the method, the ByteBuf will be positioned at the start of the first BSON object in the sequence.
+    private String getPayloadName(final ByteBuf outputByteBuf) {
+        List<Byte> payloadNameBytes = new ArrayList<>();
```

Comment:
Is there a cleaner way to do this?  I couldn't think of one.

Reply:
The issue is that `ByteBuffer`, like `byte[]`, doesn't grow.  You have to know the capacity ahead of time.

Another approach would be to save the position, loop until the zero byte to find out how big an array is needed, position the buffer back to the starting position, and fill the byte array.

Reply:
Yeah. Seems no silver bullet without knowing the size beforehand. Guava library has a Bytes.toArray util method for the last step but saw it is not in classpath. The stream usage might be much slower than vanilla loop.

Reply:
Your second approach makes lots of sense. If we could reposition to previous location, why the class didn't expose a method to provide the size?

Reply:
I think the most straightforward way is ByteArrayOutputStream: https://docs.oracle.com/javase/6/docs/api/java/io/ByteArrayOutputStream.html

Reply:
Out of curiosity, I compared the per for the following three approaches:

- approach in this PR
- approach same as in PR except that using vanilla loop instead of stream usage
- approach based on ByteArrayOutputStream usage

Below is the time spent for a huge byte stream input:

- approach #1: 93408 seconds
- apporach #2: 20667 seconds
- approach #3: 11342 seconds

It is no wonder that stream usage is 4 or 5 times slower than normal loop; ByteArrayOutputStream usage is fastest.

Reply:
As a digression, typically stream usage is 5 times slower than pre-jdk8 way. It is used in non-critical path to enhance readability (another pro is parralelStream() to tap into multi core CPU, though seldom used). As an extreme example, stream usage is only allowed in testing case in Hibernate library. It is always a good idea to inspect whether its usage is really necessary in face of its serious perf overhead, especially in library development.

Reply:
Resolving based on Valentin's suggestion below, which has been committed.